# **PANDAS**

- Data Frame
- Data Structure

## Create DataFrame using Python Dictionary

In [ ]:
# DataFrame()

import pandas as pd
emp= {"Id":[101,102,103,104,105],
      "Name":["Deepak","Shyam","Arun","Manu","Jeena"],
      "Age":[25,22,24,20,27],
      "Salary":[25000,30000,150000,10000,30000]}

df=pd.DataFrame(emp)
df


Id    Name  Age  Salary
0  101  Deepak   25   25000
1  102   Shyam   22   30000
2  103    Arun   24  150000
3  104    Manu   20   10000
4  105   Jeena   27   30000

## Create DataFrame using python tuple list

In [ ]:
# using DataFrame()

import pandas as pd
emp_data=[(101,"Deepak",25,25000),
         (102,"Shyam",22,30000),
         (103,"Arun",24,15000),
         (104,"Manu",20,10000),
         (105,"Jeena",27,30000)]
df=pd.DataFrame(emp_data,columns=["id","name","age","salary"])
df

id    name  age  salary
0  101  Deepak   25   25000
1  102   Shyam   22   30000
2  103    Arun   24   15000
3  104    Manu   20   10000
4  105   Jeena   27   30000

## Create DataFrame using a list of dictionaries

In [ ]:
# DataFrame()
import pandas as pd
emp_data_list=[{"id":101,"name":"deepak","age":25,"salary":25000},
               {"id":102,"name":"Shyam","Age":22,"salary":30000},
               {"id":103,"name":"Arun","Age":24,"salary":150000},
               {"id":104,"name":"Manu","Age":20,"salary":10000},
               {"id":105,"name":"Jeena","Age":27,"salary":30000}]
df=pd.DataFrame(emp_data_list)
df

id    name   age  salary   Age
0  101  deepak  25.0   25000   NaN
1  102   Shyam   NaN   30000  22.0
2  103    Arun   NaN  150000  24.0
3  104    Manu   NaN   10000  20.0
4  105   Jeena   NaN   30000  27.0

## Create DataFrame using python tuple list

In [ ]:
import pandas as pd
emp_data=[(101,"Deepak",25,25000),
(102,"Shyam",22,30000),
(103,"Arun",24,150000),
(104,"Manu",20,10000),
(105,"Jeena",27,30000)]

df=pd.DataFrame(emp_data,columns=["ID","NAME","AGE","SALARY"])
df

ID    NAME  AGE  SALARY
0  101  Deepak   25   25000
1  102   Shyam   22   30000
2  103    Arun   24  150000
3  104    Manu   20   10000
4  105   Jeena   27   30000

## Create dataframe using a list of dictionaries

In [ ]:
import pandas as pd
emp_data_list=[{"ID":101,"NAME":"DEEPAK","AGE":25,"SALARY":25000},
              {"ID":102,"NAME":"SHYAM","AGE":22,"SALARY":30000},
              {"ID":103,"NAME":"ARUN","AGE":24,"SALARY":150000},
              {"ID":104,"NAME":"MANU","AGE":20,"SALARY":10000},
              {"ID":105,"NAME":"JEENA","AGE":27,"SALARY":30000}]
df=pd.DataFrame(emp_data_list)
df

ID    NAME  AGE  SALARY
0  101  DEEPAK   25   25000
1  102   SHYAM   22   30000
2  103    ARUN   24  150000
3  104    MANU   20   10000
4  105   JEENA   27   30000

# Create DataFrame using CSV

file datakal table aakan use cheyyunnuuu..

In [ ]:
dt = pd.read_csv("/content/sample_data/mnist_train_small.csv")
dt

6  0  0.1  0.2  0.3  0.4  0.5  0.6  0.7  0.8  ...  0.581  0.582  0.583  \
0      5  0    0    0    0    0    0    0    0    0  ...      0      0      0   
1      7  0    0    0    0    0    0    0    0    0  ...      0      0      0   
2      9  0    0    0    0    0    0    0    0    0  ...      0      0      0   
3      5  0    0    0    0    0    0    0    0    0  ...      0      0      0   
4      2  0    0    0    0    0    0    0    0    0  ...      0      0      0   
...   .. ..  ...  ...  ...  ...  ...  ...  ...  ...  ...    ...    ...    ...   
19994  0  0    0    0    0    0    0    0    0    0  ...      0      0      0   
19995  1  0    0    0    0    0    0    0    0    0  ...      0      0      0   
19996  2  0    0    0    0    0    0    0    0    0  ...      0      0      0   
19997  9  0    0    0    0    0    0    0    0    0  ...      0      0      0   
19998  5  0    0    0    0    0    0    0    0    0  ...      0      0      0   

       0.584  0.585  0.586  0.587  0.588  0.589  0.590  
0          0      0      0      0      0      0      0  
1          0      0      0      0      0      0      0  
2          0      0      0      0      0      0      0  
3          0      0      0      0      0      0      0  
4          0      0      0      0      0      0      0  
...      ...    ...    ...    ...    ...    ...    ...  
19994      0      0      0      0      0      0      0  
19995      0      0      0      0      0      0      0  
19996      0      0      0      0      0      0      0  
19997      0      0      0      0      0      0      0  
19998      0      0      0      0      0      0      0  

[19999 rows x 785 columns]

Files l - sample data-csv files

In [ ]:
data=pd.read_csv("/content/sample_data/california_housing_train.csv")
dt

6  0  0.1  0.2  0.3  0.4  0.5  0.6  0.7  0.8  ...  0.581  0.582  0.583  \
0      5  0    0    0    0    0    0    0    0    0  ...      0      0      0   
1      7  0    0    0    0    0    0    0    0    0  ...      0      0      0   
2      9  0    0    0    0    0    0    0    0    0  ...      0      0      0   
3      5  0    0    0    0    0    0    0    0    0  ...      0      0      0   
4      2  0    0    0    0    0    0    0    0    0  ...      0      0      0   
...   .. ..  ...  ...  ...  ...  ...  ...  ...  ...  ...    ...    ...    ...   
19994  0  0    0    0    0    0    0    0    0    0  ...      0      0      0   
19995  1  0    0    0    0    0    0    0    0    0  ...      0      0      0   
19996  2  0    0    0    0    0    0    0    0    0  ...      0      0      0   
19997  9  0    0    0    0    0    0    0    0    0  ...      0      0      0   
19998  5  0    0    0    0    0    0    0    0    0  ...      0      0      0   

       0.584  0.585  0.586  0.587  0.588  0.589  0.590  
0          0      0      0      0      0      0      0  
1          0      0      0      0      0      0      0  
2          0      0      0      0      0      0      0  
3          0      0      0      0      0      0      0  
4          0      0      0      0      0      0      0  
...      ...    ...    ...    ...    ...    ...    ...  
19994      0      0      0      0      0      0      0  
19995      0      0      0      0      0      0      0  
19996      0      0      0      0      0      0      0  
19997      0      0      0      0      0      0      0  
19998      0      0      0      0      0      0      0  

[19999 rows x 785 columns]

# Pandas Series

In [ ]:
# create a sample pandas series from a list
import pandas as pd
a=[1,7,2]
myvar=pd.Series(a)
print(myvar)


0    1
1    7
2    2
dtype: int64


# Labels

In [ ]:
# This label can be used to access a specified value

import pandas as pd
a =[1,7,2]
myvar=pd.Series(a)
print(myvar[0])

1


## Create Labels

In [ ]:
# index argument

import pandas as pd
a=[1,7,2]
myvar=pd.Series(a,index=["x","y","z"])
print(myvar)

x    1
y    7
z    2
dtype: int64


In [ ]:
import pandas as pd
a=[1,7,2]
myvar=pd.Series(a,index=["x","y","z"])
print(myvar["y"])

7


##  Key/value objects as series

In [ ]:
import pandas as pd
calories = {"day1":420,"day2":380,"day3":390}
myvar=pd.Series(calories)
print(myvar)

day1    420
day2    380
day3    390
dtype: int64


In [ ]:
import pandas as pd
calories={"day1":420,"day2":380,"day3":390}
myvar=pd.Series(calories,index=["day1","day2"])
print(myvar)

day1    420
day2    380
dtype: int64


# Pandas Read CSV

In [ ]:
import pandas as pd
data=pd.read_csv("/content/drive/MyDrive/Colab Notebooks/data.csv")
data

Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
..        ...    ...       ...       ...
164        60    105       140     290.8
165        60    110       145     300.0
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4

[169 rows x 4 columns]

In [ ]:
import pandas as pd
data=pd.read_csv("/content/drive/MyDrive/Colab Notebooks/data.csv")
print(data.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

## Max_Rows

In [ ]:
# check the number of maximum returned rows
import pandas as pd
data=pd.read_csv("/content/drive/MyDrive/Colab Notebooks/data.csv")
print(pd.options.display.max_rows)

60


increase the maximum number of rows to display the entire dataframe

In [ ]:
import pandas as pd
pd.options.display.max_rows=9999
df=pd.read_csv("/content/drive/MyDrive/Colab Notebooks/data.csv")
print(df)

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

# Pandas Read JSON

In [ ]:
import pandas as pd
df = pd.read_json("/content/drive/MyDrive/Colab Notebooks/data.json")
print(df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.5
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

## Dictionary as JSON

In [ ]:
import pandas as pd
data = {
  "Duration":{
    "0":60,
    "1":60,
    "2":60,
    "3":45,
    "4":45,
    "5":60
  },
  "Pulse":{
    "0":110,
    "1":117,
    "2":103,
    "3":109,
    "4":117,
    "5":102
  },
  "Maxpulse":{
    "0":130,
    "1":145,
    "2":135,
    "3":175,
    "4":148,
    "5":127
  },
  "Calories":{
    "0":409,
    "1":479,
    "2":340,
    "3":282,
    "4":406,
    "5":300
  }
}
df=pd.DataFrame(data)
print(df)

   Duration  Pulse  Maxpulse  Calories
0        60    110       130       409
1        60    117       145       479
2        60    103       135       340
3        45    109       175       282
4        45    117       148       406
5        60    102       127       300


In [ ]:
import pandas as pd

data  = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/data_frame.csv")
data

Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.5
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45    105       132     246.0
24         60    102       126     334.5
25         60    100       120     250.0
26         60     92       118     241.0
27         60    103       132       NaN
28         60    100       132     280.0
29         60    102       129     380.3
30         60     92       115     243.0
31         45     90       112     180.1
32         60    101       124     299.0
33         60     93       113     223.0
34         60    107       136     361.0
35         60    114       140     415.0
36         60    102       127     300.5
37         60    100       120     300.1
38         60    100       120     300.0
39         45    104       129     266.0
40         45     90       112     180.1
41         60     98       126     286.0
42         60    100       122     329.4
43         60    111       138     400.0
44         60    111       131     397.0
45         60     99       119     273.0
46         60    109       153     387.6
47         45    111       136     300.0
48         45    108       129     298.0
49         60    111       139     397.6
50         60    107       136     380.2
51         80    123       146     643.1
52         60    106       130     263.0
53         60    118       151     486.0
54         30    136       175     238.0
55         60    121       146     450.7
56         60    118       121     413.0
57         45    115       144     305.0
58         20    153       172     226.4
59         45    123       152     321.0
60        210    108       160    1376.0
61        160    110       137    1034.4
62        160    109       135     853.0
63         45    118       141     341.0
64         20    110       130     131.4
65        180     90       130     800.4
66        150    105       135     873.4
67        150    107       130     816.0
68         20    106       136     110.4
69        300    108       143    1500.2
70        150     97       129    1115.0
71         60    109       153     387.6
72         90    100       127     700.0
73        150     97       127     953.2
74         45    114       146     304.0
75         90     98       125     563.2
76         45    105       134     251.0
77         45    110       141     300.0
78        120    100       130     500.4
79        270    100       131    1729.0
80         30    159       182     319.2
81         45    149       169     344.0
82         30    103       139     151.1
83        120    100       130     500.0
84         45    100       120     225.3
85         30    151       170     300.1
86         45    102       136     234.0
87        120    100       157    1000.1
88         45    129       103     242.0
89         20     83       107      50.3
90        180    101       127     600.1
91         45    107       137       NaN
92         30     90       107     105.3
93         15     80       100      50.5
94         20    150       171     127.4
95         20    151       168     229.4
96         30     95       1

In [ ]:
data.head(4)

Duration  Pulse  Maxpulse  Calories
0        60    110       130     409.1
1        60    117       145     479.0
2        60    103       135     340.0
3        45    109       175     282.4

In [ ]:
data.tail(10)

Duration  Pulse  Maxpulse  Calories
159        30     80       120     240.9
160        30     85       120     250.4
161        45     90       130     260.4
162        45     95       130     270.0
163        45    100       140     280.9
164        60    105       140     290.8
165        60    110       145     300.4
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 169 entries, 0 to 168
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Duration  169 non-null    int64  
 1   Pulse     169 non-null    int64  
 2   Maxpulse  169 non-null    int64  
 3   Calories  164 non-null    float64
dtypes: float64(1), int64(3)
memory usage: 5.4 KB


In [ ]:
data.describe()

Duration       Pulse    Maxpulse     Calories
count  169.000000  169.000000  169.000000   164.000000
mean    63.846154  107.461538  134.047337   375.800000
std     42.299949   14.510259   16.450434   266.377134
min     15.000000   80.000000  100.000000    50.300000
25%     45.000000  100.000000  124.000000   250.925000
50%     60.000000  105.000000  131.000000   318.600000
75%     60.000000  111.000000  141.000000   387.600000
max    300.000000  159.000000  184.000000  1860.400000

In [ ]:
df=pd.read_csv("/content/drive/MyDrive/Colab Notebooks/data_frame.csv")
df

Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.5
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45    105       132     246.0
24         60    102       126     334.5
25         60    100       120     250.0
26         60     92       118     241.0
27         60    103       132       NaN
28         60    100       132     280.0
29         60    102       129     380.3
30         60     92       115     243.0
31         45     90       112     180.1
32         60    101       124     299.0
33         60     93       113     223.0
34         60    107       136     361.0
35         60    114       140     415.0
36         60    102       127     300.5
37         60    100       120     300.1
38         60    100       120     300.0
39         45    104       129     266.0
40         45     90       112     180.1
41         60     98       126     286.0
42         60    100       122     329.4
43         60    111       138     400.0
44         60    111       131     397.0
45         60     99       119     273.0
46         60    109       153     387.6
47         45    111       136     300.0
48         45    108       129     298.0
49         60    111       139     397.6
50         60    107       136     380.2
51         80    123       146     643.1
52         60    106       130     263.0
53         60    118       151     486.0
54         30    136       175     238.0
55         60    121       146     450.7
56         60    118       121     413.0
57         45    115       144     305.0
58         20    153       172     226.4
59         45    123       152     321.0
60        210    108       160    1376.0
61        160    110       137    1034.4
62        160    109       135     853.0
63         45    118       141     341.0
64         20    110       130     131.4
65        180     90       130     800.4
66        150    105       135     873.4
67        150    107       130     816.0
68         20    106       136     110.4
69        300    108       143    1500.2
70        150     97       129    1115.0
71         60    109       153     387.6
72         90    100       127     700.0
73        150     97       127     953.2
74         45    114       146     304.0
75         90     98       125     563.2
76         45    105       134     251.0
77         45    110       141     300.0
78        120    100       130     500.4
79        270    100       131    1729.0
80         30    159       182     319.2
81         45    149       169     344.0
82         30    103       139     151.1
83        120    100       130     500.0
84         45    100       120     225.3
85         30    151       170     300.1
86         45    102       136     234.0
87        120    100       157    1000.1
88         45    129       103     242.0
89         20     83       107      50.3
90        180    101       127     600.1
91         45    107       137       NaN
92         30     90       107     105.3
93         15     80       100      50.5
94         20    150       171     127.4
95         20    151       168     229.4
96         30     95       1

In [ ]:
df.head(5)

Duration  Pulse  Maxpulse  Calories
0        60    110       130     409.1
1        60    117       145     479.0
2        60    103       135     340.0
3        45    109       175     282.4
4        45    117       148     406.0

In [ ]:
df.tail(10)

Duration  Pulse  Maxpulse  Calories
159        30     80       120     240.9
160        30     85       120     250.4
161        45     90       130     260.4
162        45     95       130     270.0
163        45    100       140     280.9
164        60    105       140     290.8
165        60    110       145     300.4
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 169 entries, 0 to 168
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Duration  169 non-null    int64  
 1   Pulse     169 non-null    int64  
 2   Maxpulse  169 non-null    int64  
 3   Calories  164 non-null    float64
dtypes: float64(1), int64(3)
memory usage: 5.4 KB


In [ ]:
df.describe()

Duration       Pulse    Maxpulse     Calories
count  169.000000  169.000000  169.000000   164.000000
mean    63.846154  107.461538  134.047337   375.800000
std     42.299949   14.510259   16.450434   266.377134
min     15.000000   80.000000  100.000000    50.300000
25%     45.000000  100.000000  124.000000   250.925000
50%     60.000000  105.000000  131.000000   318.600000
75%     60.000000  111.000000  141.000000   387.600000
max    300.000000  159.000000  184.000000  1860.400000

# Cleaning Empty Cells

In [ ]:
# Return a new DataFrame with no empty cell

import pandas as pd
df = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/data_frame.csv")
new_df = df.dropna()
print(new_df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.5
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45    105       132     246.0
24         60   

In [ ]:
# Remove all rows with NULL values

import pandas as pd
df = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/data_frame.csv")
df.dropna(inplace=True)
print(df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.5
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45    105       132     246.0
24         60   

## Replace Empty Values

In [ ]:
# Replace NULL values with the number 130

import pandas as pd
df = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/data_frame.csv")
df.fillna(130,inplace =True)
print(df.to_string)

<bound method DataFrame.to_string of      Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.5
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112     130.0
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130

In [ ]:
# Replace NULL values in the "Calories" columns with the number 130
import pandas as pd
df = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/data_frame.csv")
df.fillna({"Calories": 130},inplace = True)
print(df.to_string)

<bound method DataFrame.to_string of      Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.5
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112     130.0
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130

# Replace Using Mean,Median , or Mode

- A common way to replace empty cells,is to calculate the mean ,median,or mode value of the column.
- Pandas uses the mean(),median(),and mode() methods to calculate the respective values for a specified column

In [ ]:
# Calculate the MEAN , and replace any empty values with it

import pandas as pd
df = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/data_frame.csv")
x = df["Calories"].mean()
df.fillna({"Calories": x},inplace =True)
print(df.to_string)

<bound method DataFrame.to_string of      Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.5
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112     375.8
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130

In [ ]:
# Calculate the MEDIAN , and replace any empty values with it

import pandas as pd
df = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/data_frame.csv")
x=df["Calories"].median()
df.fillna({"Calories": x},inplace =True)
print(df.to_string)

<bound method DataFrame.to_string of      Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.5
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112     318.6
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130

In [ ]:
# Calculate the MODE ,and replace any empty values with it

import pandas as pd
df = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/data_frame.csv")
x=df["Calories"].mode()[0]
df.fillna({"Calories": x},inplace = True)
print(df.to_string)

<bound method DataFrame.to_string of      Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.5
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112     300.0
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130

# Cleaning Data of Wrong Format

In [ ]:
import pandas as pd
data = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/calories.csv")
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32 entries, 0 to 31
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Duration  32 non-null     int64  
 1   Date      31 non-null     object 
 2   Pulse     32 non-null     int64  
 3   Maxpulse  32 non-null     int64  
 4   Calories  30 non-null     float64
dtypes: float64(1), int64(3), object(1)
memory usage: 1.4+ KB


In [ ]:
data

Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'    100       120     300.0
18        45  '2020/12/18'     90       112       NaN
19        60  '2020/12/19'    103       123     323.0
20        45  '2020/12/20'     97       125     243.0
21        60  '2020/12/21'    108       131     364.2
22        45           NaN    100       119     282.0
23        60  '2020/12/23'    130       101     300.0
24        45  '2020/12/24'    105       132     246.0
25        60  '2020/12/25'    102       126     334.5
26        60      20201226    100       120     250.0
27        60  '2020/12/27'     92       118     241.0
28        60  '2020/12/28'    103       132       NaN
29        60  '2020/12/29'    100       132     280.0
30        60  '2020/12/30'    102       129     380.3
31        60  '2020/12/31'     92       115     243.0

In [ ]:
data['Date']= pd.to_datetime(data['Date'],format='mixed')

In [ ]:
data

Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7        450 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90       112       NaN
19        60 2020-12-19    103       123     323.0
20        45 2020-12-20     97       125     243.0
21        60 2020-12-21    108       131     364.2
22        45        NaT    100       119     282.0
23        60 2020-12-23    130       101     300.0
24        45 2020-12-24    105       132     246.0
25        60 2020-12-25    102       126     334.5
26        60 2020-12-26    100       120     250.0
27        60 2020-12-27     92       118     241.0
28        60 2020-12-28    103       132       NaN
29        60 2020-12-29    100       132     280.0
30        60 2020-12-30    102       129     380.3
31        60 2020-12-31     92       115     243.0

In [ ]:
data.dropna(subset=['Date'],inplace=True)

In [ ]:
data

Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7        450 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90       112       NaN
19        60 2020-12-19    103       123     323.0
20        45 2020-12-20     97       125     243.0
21        60 2020-12-21    108       131     364.2
23        60 2020-12-23    130       101     300.0
24        45 2020-12-24    105       132     246.0
25        60 2020-12-25    102       126     334.5
26        60 2020-12-26    100       120     250.0
27        60 2020-12-27     92       118     241.0
28        60 2020-12-28    103       132       NaN
29        60 2020-12-29    100       132     280.0
30        60 2020-12-30    102       129     380.3
31        60 2020-12-31     92       115     243.0

## Convert Into a Correct Format

In [ ]:
# convert to date
import pandas as pd
df = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/calories.csv")
df["Date"] = pd.to_datetime(df["Date"],format = "mixed")
print(df.to_string())

    Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7        450 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90 

In [ ]:
df.dropna(subset=["Date"],inplace=True)
data

Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7        450 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90       112       NaN
19        60 2020-12-19    103       123     323.0
20        45 2020-12-20     97       125     243.0
21        60 2020-12-21    108       131     364.2
23        60 2020-12-23    130       101     300.0
24        45 2020-12-24    105       132     246.0
25        60 2020-12-25    102       126     334.5
26        60 2020-12-26    100       120     250.0
27        60 2020-12-27     92       118     241.0
28        60 2020-12-28    103       132       NaN
29        60 2020-12-29    100       132     280.0
30        60 2020-12-30    102       129     380.3
31        60 2020-12-31     92       115     243.0

# Fixing Wrong Data

- Wrong Data
"Wrong data" does not have to be "empty cells" or "wrong format", it can just be wrong, like if someone registered "199" instead of "1.99".

- Sometimes you can spot wrong data by looking at the data set, because you have an expectation of what it should be.

- If you take a look at our data set, you can see that in row 7, the duration is 450, but for all the other rows the duration is between 30 and 60.



## Removing Duplicates

In [ ]:
print(df.duplicated())

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12     True
13    False
14    False
15    False
16    False
17    False
19    False
20    False
21    False
23    False
24    False
25    False
26    False
27    False
29    False
30    False
31    False
dtype: bool


In [ ]:
import pandas as pd
df = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/calories.csv")
df.drop_duplicates(inplace = True)
print(df.to_string())
# 12 has been removed from the result

    Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'    100       120     300.0
18        45  '2020/12/18'  

## Replacing Values

In [ ]:
# set "duration" =45 in row 7

df.loc[7,"Duration"]=45
df

Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7         45  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'    100       120     300.0
18        45  '2020/12/18'     90       112       NaN
19        60  '2020/12/19'    103       123     323.0
20        45  '2020/12/20'     97       125     243.0
21        60  '2020/12/21'    108       131     364.2
22        45           NaN    100       119     282.0
23        60  '2020/12/23'    130       101     300.0
24        45  '2020/12/24'    105       132     246.0
25        60  '2020/12/25'    102       126     334.5
26        60      20201226    100       120     250.0
27        60  '2020/12/27'     92       118     241.0
28        60  '2020/12/28'    103       132       NaN
29        60  '2020/12/29'    100       132     280.0
30        60  '2020/12/30'    102       129     380.3
31        60  '2020/12/31'     92       115     243.0

In [ ]:
# loop through all values in the "Duration" column.
# if the value is higher than 120,set it to 120

for x in df.index:
  if df.loc[x,"Duration"] > 120:
    df.loc[x,"Duration"]= 120
df

Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7         45  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'    100       120     300.0
18        45  '2020/12/18'     90       112       NaN
19        60  '2020/12/19'    103       123     323.0
20        45  '2020/12/20'     97       125     243.0
21        60  '2020/12/21'    108       131     364.2
22        45           NaN    100       119     282.0
23        60  '2020/12/23'    130       101     300.0
24        45  '2020/12/24'    105       132     246.0
25        60  '2020/12/25'    102       126     334.5
26        60      20201226    100       120     250.0
27        60  '2020/12/27'     92       118     241.0
28        60  '2020/12/28'    103       132       NaN
29        60  '2020/12/29'    100       132     280.0
30        60  '2020/12/30'    102       129     380.3
31        60  '2020/12/31'     92       115     243.0

## Removing Rows

In [ ]:
# Delete rows where "Duration" is higher than 120

for x in df.index:
  if df.loc[x,"Duration"] > 120:
    df.drop(x,inplace =True)
df

Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7         45  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'    100       120     300.0
18        45  '2020/12/18'     90       112       NaN
19        60  '2020/12/19'    103       123     323.0
20        45  '2020/12/20'     97       125     243.0
21        60  '2020/12/21'    108       131     364.2
22        45           NaN    100       119     282.0
23        60  '2020/12/23'    130       101     300.0
24        45  '2020/12/24'    105       132     246.0
25        60  '2020/12/25'    102       126     334.5
26        60      20201226    100       120     250.0
27        60  '2020/12/27'     92       118     241.0
28        60  '2020/12/28'    103       132       NaN
29        60  '2020/12/29'    100       132     280.0
30        60  '2020/12/30'    102       129     380.3
31        60  '2020/12/31'     92       115     243.0

# Data Correlations

- A great aspect of the Pandas module is the corr() method.

- The corr() method calculates the relationship between each column in your data set.

In [ ]:
import pandas as pd
df = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/calories.csv")
df.corr

<bound method DataFrame.corr of     Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'    100       120     300.0
18        45  '2020/12/18'     90       112       NaN
19        60  '2020/12/19'    103       123     323.0
20        45  '2020/12/20'     97       125     243.0
21        60  '2020/12/21'    108       131     364.2
22        45           NaN    100       119     282.0
23        60  '2020/12/23'    130       101     300.0
24        45  '2020/12/24'    105       132     246.0
25        60  '2020/12/25'    102       126     334.5
26        60      20201226    100       120     250.0
27        60  '2020/12/27'     92       118     241.0
28        60  '2020/12/28'    103       132       NaN
29        60  '2020/12/29'    100       132     280.0
30        60  '2020/12/30'    102       129     380.3
31        60  '2020/12/31'     92       115     243.0>

## to_datetime()

to_datetime function is very useful in pandas to convert any string date to type datetime

In [ ]:
import pandas as pd
pd.to_datetime(["05/01/2023"],dayfirst = True)

DatetimeIndex(['2023-01-05'], dtype='datetime64[ns]', freq=None)

In [ ]:
dates = ["2023-01-01","invalid_date","2023-03-30"]
pd.to_datetime(dates,errors="coerce")

DatetimeIndex(['2023-01-01', 'NaT', '2023-03-30'], dtype='datetime64[ns]', freq=None)

In [ ]:
import pandas as pd
df = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/calories.csv")
df

Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'    100       120     300.0
18        45  '2020/12/18'     90       112       NaN
19        60  '2020/12/19'    103       123     323.0
20        45  '2020/12/20'     97       125     243.0
21        60  '2020/12/21'    108       131     364.2
22        45           NaN    100       119     282.0
23        60  '2020/12/23'    130       101     300.0
24        45  '2020/12/24'    105       132     246.0
25        60  '2020/12/25'    102       126     334.5
26        60      20201226    100       120     250.0
27        60  '2020/12/27'     92       118     241.0
28        60  '2020/12/28'    103       132       NaN
29        60  '2020/12/29'    100       132     280.0
30        60  '2020/12/30'    102       129     380.3
31        60  '2020/12/31'     92       115     243.0

## Using Converters

Using converters to convert data using conditions

In [ ]:
import pandas as pd
df=pd.read_csv("/content/drive/MyDrive/Colab Notebooks/data (2).csv")
print(df)

    name                age occupation
0   john                 23   Engineer
1  alice                 30     doctor
2   ammu  not available age    teacher


In [ ]:
# Convert occupation of doctor to teacher
import pandas as pd
def convert(occupation):
  if occupation=="doctor":
    return "teacher"
  else:
    return occupation
df=pd.read_csv("/content/drive/MyDrive/Colab Notebooks/data (2).csv",converters={"occupation":convert})
print(df)

    name                age occupation
0   john                 23   Engineer
1  alice                 30    teacher
2   ammu  not available age    teacher


# Group by

In [ ]:
# group by

group_result=df.groupby("airline") # group by airplines
list(group_result)

[('AirAsia',
           index  airline   flight source_city departure_time stops  \
  2            2  AirAsia   I5-764       Delhi  Early_Morning  zero   
  18          18  AirAsia   I5-747       Delhi        Evening   one   
  19          19  AirAsia   I5-747       Delhi        Evening   one   
  27          27  AirAsia   I5-744       Delhi        Morning   one   
  48          48  AirAsia   I5-784       Delhi        Evening   one   
  ...        ...      ...      ...         ...            ...   ...   
  206538  206538  AirAsia  I5-1229     Chennai          Night   one   
  206539  206539  AirAsia  I5-2462     Chennai          Night   one   
  206591  206591  AirAsia   I5-517     Chennai        Morning  zero   
  206616  206616  AirAsia  I5-1229     Chennai          Night   one   
  206617  206617  AirAsia  I5-2462     Chennai          Night   one   
  
           arrival_time destination_city    class  duration  days_left  price  
  2       Early_Morning           Mumbai  Economy   

In [ ]:
import pandas as pd
df=pd.read_csv("/content/drive/MyDrive/Colab Notebooks/airlines_flights_data.csv")# group by airlines
df

index   airline   flight source_city departure_time stops  \
0            0  SpiceJet  SG-8709       Delhi        Evening  zero   
1            1  SpiceJet  SG-8157       Delhi  Early_Morning  zero   
2            2   AirAsia   I5-764       Delhi  Early_Morning  zero   
3            3   Vistara   UK-995       Delhi        Morning  zero   
4            4   Vistara   UK-963       Delhi        Morning  zero   
...        ...       ...      ...         ...            ...   ...   
300148  300148   Vistara   UK-822     Chennai        Morning   one   
300149  300149   Vistara   UK-826     Chennai      Afternoon   one   
300150  300150   Vistara   UK-832     Chennai  Early_Morning   one   
300151  300151   Vistara   UK-828     Chennai  Early_Morning   one   
300152  300152   Vistara   UK-822     Chennai        Morning   one   

         arrival_time destination_city     class  duration  days_left  price  
0               Night           Mumbai   Economy      2.17          1   5953  
1             Morning           Mumbai   Economy      2.33          1   5953  
2       Early_Morning           Mumbai   Economy      2.17          1   5956  
3           Afternoon           Mumbai   Economy      2.25          1   5955  
4             Morning           Mumbai   Economy      2.33          1   5955  
...               ...              ...       ...       ...        ...    ...  
300148        Evening        Hyderabad  Business     10.08         49  69265  
300149          Night        Hyderabad  Business     10.42         49  77105  
300150          Night        Hyderabad  Business     13.83         49  79099  
300151        Evening        Hyderabad  Business     10.00         49  81585  
300152        Evening        Hyderabad  Business     10.08         49  81585  

[300153 rows x 12 columns]

In [ ]:
df.groupby("destination_city")["price"].max()

destination_city
Bangalore    114523
Chennai      111964
Delhi        123071
Hyderabad    114705
Kolkata      117307
Mumbai       115211
Name: price, dtype: int64

In [ ]:
df.groupby("source_city")["duration"].min()

source_city
Bangalore    0.83
Chennai      0.83
Delhi        1.92
Hyderabad    0.92
Kolkata      2.08
Mumbai       1.17
Name: duration, dtype: float64

## Concat

- Concat() function is used to concatenate 2 or more dataframes.
- Contact can do both vertically(axis=0) as well as horizontal(axis=1)

In [ ]:
tcs = {"Id":[101,102,103,104],
       "Name":["Mithun","Dipin","Jose","Rahul"],
       "Age":[25,26,27,29]}

wipro = {"Id":[101,102,103,104],
        "Name":["Dev","Sreenath","Sreerag","Shafeel"],
        "Age":[24,24,27,26]}

tcs_emp = pd.DataFrame(tcs)
wipro_emp = pd.DataFrame(wipro)
pd.concat([tcs_emp,wipro_emp])

Id      Name  Age
0  101    Mithun   25
1  102     Dipin   26
2  103      Jose   27
3  104     Rahul   29
0  101       Dev   24
1  102  Sreenath   24
2  103   Sreerag   27
3  104   Shafeel   26

In [ ]:
# ignore index - To get common index after concatenation

pd.concat([tcs_emp,wipro_emp],ignore_index=True)


Id      Name  Age
0  101    Mithun   25
1  102     Dipin   26
2  103      Jose   27
3  104     Rahul   29
4  101       Dev   24
5  102  Sreenath   24
6  103   Sreerag   27
7  104   Shafeel   26

In [ ]:
# Keys - To provide seperate keys for each concatenate data frames
data=pd.concat([tcs_emp,wipro_emp],keys=["TCS","WIPRO"])
data

Id      Name  Age
TCS   0  101    Mithun   25
      1  102     Dipin   26
      2  103      Jose   27
      3  104     Rahul   29
WIPRO 0  101       Dev   24
      1  102  Sreenath   24
      2  103   Sreerag   27
      3  104   Shafeel   26

In [ ]:
# Concatenating vertically
data=pd.concat([tcs_emp,wipro_emp],axis=1)
data

Id    Name  Age   Id      Name  Age
0  101  Mithun   25  101       Dev   24
1  102   Dipin   26  102  Sreenath   24
2  103    Jose   27  103   Sreerag   27
3  104   Rahul   29  104   Shafeel   26

In [ ]:
# Concatenating a dataframe with a series

salary=pd.Series([65000,75000,15000,25000],name="salary")
pd.concat([tcs_emp,salary],axis=1)

Id    Name  Age  salary
0  101  Mithun   25   65000
1  102   Dipin   26   75000
2  103    Jose   27   15000
3  104   Rahul   29   25000

## Merge

Merge is a process to merge 2 table horizontally

In [ ]:
tcs = {"Id":[101,102,103,104],
       "Name":["Mithun","Dipin","Jose","Rahul"],
       "Age":[25,26,27,29]}
wipro = {"Id":[101,102,103,104],
        "Name":["Dev","Sreenath","Sreerag","Shafeel"],
        "Age":[24,24,27,26]}

tcs_emp = pd.DataFrame(tcs)
wipro_emp = pd.DataFrame(wipro)
pd.concat([tcs_emp,wipro_emp])
tcs_emp

Id    Name  Age
0  101  Mithun   25
1  102   Dipin   26
2  103    Jose   27
3  104   Rahul   29

In [ ]:
wipro_emp=pd.DataFrame(wipro)
wipro_emp

Id      Name  Age
0  101       Dev   24
1  102  Sreenath   24
2  103   Sreerag   27
3  104   Shafeel   26

In [ ]:
pd.merge(tcs_emp,wipro_emp,on="Id")

Id  Name_x  Age_x    Name_y  Age_y
0  101  Mithun     25       Dev     24
1  102   Dipin     26  Sreenath     24
2  103    Jose     27   Sreerag     27
3  104   Rahul     29   Shafeel     26

In [ ]:
pd.merge(tcs_emp,wipro_emp,on="Id",suffixes=["left","right"])

Id Nameleft  Ageleft Nameright  Ageright
0  101   Mithun       25       Dev        24
1  102    Dipin       26  Sreenath        24
2  103     Jose       27   Sreerag        27
3  104    Rahul       29   Shafeel        26

In [ ]:
pd.merge(tcs_emp,wipro_emp,on="Id",how="outer") # outer join

Id  Name_x  Age_x    Name_y  Age_y
0  101  Mithun     25       Dev     24
1  102   Dipin     26  Sreenath     24
2  103    Jose     27   Sreerag     27
3  104   Rahul     29   Shafeel     26

In [ ]:
pd.merge(tcs_emp,wipro_emp,on="Id",how="outer",indicator=True)

Id  Name_x  Age_x    Name_y  Age_y _merge
0  101  Mithun     25       Dev     24   both
1  102   Dipin     26  Sreenath     24   both
2  103    Jose     27   Sreerag     27   both
3  104   Rahul     29   Shafeel     26   both

In [ ]:
pd.merge(tcs_emp,wipro_emp,on="Id",how="left")  # left join

Id  Name_x  Age_x    Name_y  Age_y
0  101  Mithun     25       Dev     24
1  102   Dipin     26  Sreenath     24
2  103    Jose     27   Sreerag     27
3  104   Rahul     29   Shafeel     26

In [ ]:
pd.merge(tcs_emp,wipro_emp,on="Id",how="right")  # right join

Id  Name_x  Age_x    Name_y  Age_y
0  101  Mithun     25       Dev     24
1  102   Dipin     26  Sreenath     24
2  103    Jose     27   Sreerag     27
3  104   Rahul     29   Shafeel     26

# Pivot and pivot tables

- Pivot allows to reshape the data frame
- Pivot table is used to summarize or aggregate data with in a dataframe.


In [ ]:
import pandas as pd
datas=pd.read_csv("/content/drive/MyDrive/weather_dataset.csv")
datas

Date       City  Temperature  Humidity  Wind Speed (km/h)  \
0  2025-08-01     Mumbai           30        85                 12   
1  2025-08-01      Delhi           36        40                 18   
2  2025-08-01  Bengaluru           27        70                 10   
3  2025-08-02     Mumbai           29        88                 14   
4  2025-08-02      Delhi           35        45                 20   
5  2025-08-02  Bengaluru           26        72                  9   
6  2025-08-03     Mumbai           28        90                 11   
7  2025-08-03      Delhi           34        50                 22   
8  2025-08-03  Bengaluru           27        68                  8   

  Weather Condition  
0             Rainy  
1             Sunny  
2            Cloudy  
3             Rainy  
4             Sunny  
5            Cloudy  
6             Rainy  
7     Partly Cloudy  
8            Cloudy

In [ ]:
datas.pivot(index="Date",columns="City") # create a table with date as index and city as columns


Temperature               Humidity              Wind Speed (km/h)  \
City         Bengaluru Delhi Mumbai Bengaluru Delhi Mumbai         Bengaluru   
Date                                                                           
2025-08-01          27    36     30        70    40     85                10   
2025-08-02          26    35     29        72    45     88                 9   
2025-08-03          27    34     28        68    50     90                 8   

                        Weather Condition                        
City       Delhi Mumbai         Bengaluru          Delhi Mumbai  
Date                                                             
2025-08-01    18     12            Cloudy          Sunny  Rainy  
2025-08-02    20     14            Cloudy          Sunny  Rainy  
2025-08-03    22     11            Cloudy  Partly Cloudy  Rainy

In [ ]:
datas.pivot(index="Date",columns="City",values="Temperature")

City        Bengaluru  Delhi  Mumbai
Date                                
2025-08-01         27     36      30
2025-08-02         26     35      29
2025-08-03         27     34      28

In [ ]:
datas.pivot_table(index="City",columns="Date",aggfunc="sum")

Humidity                       Temperature                        \
Date      2025-08-01 2025-08-02 2025-08-03  2025-08-01 2025-08-02 2025-08-03   
City                                                                           
Bengaluru         70         72         68          27         26         27   
Delhi             40         45         50          36         35         34   
Mumbai            85         88         90          30         29         28   

          Weather Condition                           Wind Speed (km/h)  \
Date             2025-08-01 2025-08-02     2025-08-03        2025-08-01   
City                                                                      
Bengaluru            Cloudy     Cloudy         Cloudy                10   
Delhi                 Sunny      Sunny  Partly Cloudy                18   
Mumbai                Rainy      Rainy          Rainy                12   

                                 
Date      2025-08-02 2025-08-03  
City                             
Bengaluru          9          8  
Delhi             20         22  
Mumbai            14         11

In [ ]:
datas.pivot_table(index="City",columns="Date",aggfunc="sum",margins=True)

Humidity                            Temperature             \
Date      2025-08-01 2025-08-02 2025-08-03  All  2025-08-01 2025-08-02   
City                                                                     
Bengaluru         70         72         68  210          27         26   
Delhi             40         45         50  135          36         35   
Mumbai            85         88         90  263          30         29   
All              195        205        208  608          93         90   

                          Weather Condition                    \
Date      2025-08-03  All        2025-08-01        2025-08-02   
City                                                            
Bengaluru         27   80            Cloudy            Cloudy   
Delhi             34  105             Sunny             Sunny   
Mumbai            28   87             Rainy             Rainy   
All               89  272  RainySunnyCloudy  RainySunnyCloudy   

                                     \
Date                     2025-08-03   
City                                  
Bengaluru                    Cloudy   
Delhi                 Partly Cloudy   
Mumbai                        Rainy   
All        RainyPartly CloudyCloudy   

                                                              \
Date                                                     All   
City                                                           
Bengaluru                                 CloudyCloudyCloudy   
Delhi                                SunnySunnyPartly Cloudy   
Mumbai                                       RainyRainyRainy   
All        RainySunnyCloudyRainySunnyCloudyRainyPartly Cl...   

          Wind Speed (km/h)                             
Date             2025-08-01 2025-08-02 2025-08-03  All  
City                                                    
Bengaluru                10          9          8   27  
Delhi                    18         20         22   60  
Mumbai                   12         14         11   37  
All                      40         43         41  124

## Pivot Grouper

In [ ]:
import pandas as pd

# load dataset

df=pd.read_csv("/content/drive/MyDrive/Sales_data.csv")
df

OrderID        Date     Category     Product  Quantity  Price
0     1001  2023-01-15  Electronics      Laptop         2  70000
1     1002  2023-01-20    Furniture       Chair         4   1500
2     1003  2023-02-05  Electronics  Headphones         5   2000
3     1004  2023-02-18    Furniture       Table         1   5000
4     1005  2023-03-03   Stationery    Notebook        10     50
5     1006  2023-03-25  Electronics       Mouse         6    800
6     1007  2023-04-10    Furniture        Sofa         1  15000
7     1008  2023-04-15   Stationery         Pen        20     20
8     1009  2023-05-01  Electronics      Laptop         1  70000
9     1010  2023-05-12    Furniture       Chair         2   1500

In [ ]:
# Add total sales column

df["Total"] = df["Quantity"] * df["Price"]
df

OrderID        Date     Category     Product  Quantity  Price   Total
0     1001  2023-01-15  Electronics      Laptop         2  70000  140000
1     1002  2023-01-20    Furniture       Chair         4   1500    6000
2     1003  2023-02-05  Electronics  Headphones         5   2000   10000
3     1004  2023-02-18    Furniture       Table         1   5000    5000
4     1005  2023-03-03   Stationery    Notebook        10     50     500
5     1006  2023-03-25  Electronics       Mouse         6    800    4800
6     1007  2023-04-10    Furniture        Sofa         1  15000   15000
7     1008  2023-04-15   Stationery         Pen        20     20     400
8     1009  2023-05-01  Electronics      Laptop         1  70000   70000
9     1010  2023-05-12    Furniture       Chair         2   1500    3000

In [ ]:
# Create pivot table with Grouper (monthly)
df['Date'] = pd.to_datetime(df['Date'])
pivot = pd.pivot_table(
    df,
    index=pd.Grouper(key='Date', freq='M'),
    columns='Category',
    values='Total',
    aggfunc='sum'
)

pivot

/tmp/ipython-input-786410142.py:5: FutureWarning: 'M' is deprecated and will be removed in a future version, please use 'ME' instead.
  index=pd.Grouper(key='Date', freq='M'),


Category    Electronics  Furniture  Stationery
Date                                          
2023-01-31     140000.0     6000.0         NaN
2023-02-28      10000.0     5000.0         NaN
2023-03-31       4800.0        NaN       500.0
2023-04-30          NaN    15000.0       400.0
2023-05-31      70000.0     3000.0         NaN

## Melt

Melt allows to transform or reshape the dataframe

In [ ]:
import pandas as pd
datas=pd.read_csv("/content/drive/MyDrive/weather_dataset.csv")
datas

Date       City  Temperature  Humidity  Wind Speed (km/h)  \
0  2025-08-01     Mumbai           30        85                 12   
1  2025-08-01      Delhi           36        40                 18   
2  2025-08-01  Bengaluru           27        70                 10   
3  2025-08-02     Mumbai           29        88                 14   
4  2025-08-02      Delhi           35        45                 20   
5  2025-08-02  Bengaluru           26        72                  9   
6  2025-08-03     Mumbai           28        90                 11   
7  2025-08-03      Delhi           34        50                 22   
8  2025-08-03  Bengaluru           27        68                  8   

  Weather Condition  
0             Rainy  
1             Sunny  
2            Cloudy  
3             Rainy  
4             Sunny  
5            Cloudy  
6             Rainy  
7     Partly Cloudy  
8            Cloudy

In [ ]:
pd.melt(datas,id_vars=["Date"])

Date           variable          value
0   2025-08-01               City         Mumbai
1   2025-08-01               City          Delhi
2   2025-08-01               City      Bengaluru
3   2025-08-02               City         Mumbai
4   2025-08-02               City          Delhi
5   2025-08-02               City      Bengaluru
6   2025-08-03               City         Mumbai
7   2025-08-03               City          Delhi
8   2025-08-03               City      Bengaluru
9   2025-08-01        Temperature             30
10  2025-08-01        Temperature             36
11  2025-08-01        Temperature             27
12  2025-08-02        Temperature             29
13  2025-08-02        Temperature             35
14  2025-08-02        Temperature             26
15  2025-08-03        Temperature             28
16  2025-08-03        Temperature             34
17  2025-08-03        Temperature             27
18  2025-08-01           Humidity             85
19  2025-08-01           Humidity             40
20  2025-08-01           Humidity             70
21  2025-08-02           Humidity             88
22  2025-08-02           Humidity             45
23  2025-08-02           Humidity             72
24  2025-08-03           Humidity             90
25  2025-08-03           Humidity             50
26  2025-08-03           Humidity             68
27  2025-08-01  Wind Speed (km/h)             12
28  2025-08-01  Wind Speed (km/h)             18
29  2025-08-01  Wind Speed (km/h)             10
30  2025-08-02  Wind Speed (km/h)             14
31  2025-08-02  Wind Speed (km/h)             20
32  2025-08-02  Wind Speed (km/h)              9
33  2025-08-03  Wind Speed (km/h)             11
34  2025-08-03  Wind Speed (km/h)             22
35  2025-08-03  Wind Speed (km/h)              8
36  2025-08-01  Weather Condition          Rainy
37  2025-08-01  Weather Condition          Sunny
38  2025-08-01  Weather Condition         Cloudy
39  2025-08-02  Weather Condition          Rainy
40  2025-08-02  Weather Condition          Sunny
41  2025-08-02  Weather Condition         Cloudy
42  2025-08-03  Weather Condition          Rainy
43  2025-08-03  Weather Condition  Partly Cloudy
44  2025-08-03  Weather Condition         Cloudy

In [ ]:
melted= pd.melt(datas,
                id_vars=["City","Date","Weather Condition"],
                value_vars=["Temperature","Humidity"],
                var_name="Weather_Parameter",
                value_name="Reading")
melted

City        Date Weather Condition Weather_Parameter  Reading
0      Mumbai  2025-08-01             Rainy       Temperature       30
1       Delhi  2025-08-01             Sunny       Temperature       36
2   Bengaluru  2025-08-01            Cloudy       Temperature       27
3      Mumbai  2025-08-02             Rainy       Temperature       29
4       Delhi  2025-08-02             Sunny       Temperature       35
5   Bengaluru  2025-08-02            Cloudy       Temperature       26
6      Mumbai  2025-08-03             Rainy       Temperature       28
7       Delhi  2025-08-03     Partly Cloudy       Temperature       34
8   Bengaluru  2025-08-03            Cloudy       Temperature       27
9      Mumbai  2025-08-01             Rainy          Humidity       85
10      Delhi  2025-08-01             Sunny          Humidity       40
11  Bengaluru  2025-08-01            Cloudy          Humidity       70
12     Mumbai  2025-08-02             Rainy          Humidity       88
13      Delhi  2025-08-02             Sunny          Humidity       45
14  Bengaluru  2025-08-02            Cloudy          Humidity       72
15     Mumbai  2025-08-03             Rainy          Humidity       90
16      Delhi  2025-08-03     Partly Cloudy          Humidity       50
17  Bengaluru  2025-08-03            Cloudy          Humidity       68

## Cross tab or contingency table

- In pandas , **pd.crosstab()** is used to compute a cross-tabulation of two or more factors.Its often used for analyzing categorical variables,similar to a contingency table in statistics.

In [ ]:
import pandas as pd

df = pd.read_csv("/content/drive/MyDrive/survey_crosstab.csv")
df

RespondentID  Gender AgeGroup Preferred_Device
0             1    Male    18-25           Laptop
1             2  Female    26-35           Mobile
2             3    Male    18-25           Tablet
3             4  Female    18-25           Mobile
4             5    Male    26-35           Laptop
5             6  Female    18-25           Mobile
6             7    Male    36-45           Laptop
7             8  Female    26-35           Tablet
8             9    Male    26-35           Mobile
9            10  Female    36-45           Laptop

In [ ]:
# Create cross-tabulation

ct=pd.crosstab(df["Gender"],df["Preferred_Device"])
print(ct)

Preferred_Device  Laptop  Mobile  Tablet
Gender                                  
Female                 1       3       1
Male                   3       1       1


In [ ]:
# Cross-tab of AgeGroup vs Preferred_Device with totals

ct2= pd.crosstab(df["AgeGroup"],df["Preferred_Device"],margins = True,margins_name="Total")
print(ct2)

Preferred_Device  Laptop  Mobile  Tablet  Total
AgeGroup                                       
18-25                  1       2       1      4
26-35                  1       2       1      4
36-45                  2       0       0      2
Total                  4       4       2     10


## to_datetime()

to_datetime function is very useful in pandas to convert any string date to type datetime.

In [ ]:
import pandas as pd
pd.to_datetime(["05/01/2017"],dayfirst=True)

DatetimeIndex(['2017-01-05'], dtype='datetime64[ns]', freq=None)

In [ ]:
dates=["2023-01-01","invalid_date","2023-03-30"]
pd.to_datetime(dates,errors="coerce")

DatetimeIndex(['2023-01-01', 'NaT', '2023-03-30'], dtype='datetime64[ns]', freq=None)

In [ ]:
import pandas as pd
df=pd.read_csv("/content/drive/MyDrive/weather_dataset.csv")
df

Date       City  Temperature  Humidity  Wind Speed (km/h)  \
0  2025-08-01     Mumbai           30        85                 12   
1  2025-08-01      Delhi           36        40                 18   
2  2025-08-01  Bengaluru           27        70                 10   
3  2025-08-02     Mumbai           29        88                 14   
4  2025-08-02      Delhi           35        45                 20   
5  2025-08-02  Bengaluru           26        72                  9   
6  2025-08-03     Mumbai           28        90                 11   
7  2025-08-03      Delhi           34        50                 22   
8  2025-08-03  Bengaluru           27        68                  8   

  Weather Condition  
0             Rainy  
1             Sunny  
2            Cloudy  
3             Rainy  
4             Sunny  
5            Cloudy  
6             Rainy  
7     Partly Cloudy  
8            Cloudy

In [ ]:
print("Original DataFrame:")
print(df)

# use applymap() to increase all numeric values by 1.0 (if applicable)
# we must exclude the "City" column (non-numeric),or handle non-numeric safely

df_numeric = df.select_dtypes(include="number")
df[df_numeric.columns]=df_numeric.applymap(lambda x:x + 1)

print("\nDataFrame after applymap (numeric values +1):")
df

Original DataFrame:
         Date       City  Temperature  Humidity  Wind Speed (km/h)  \
0  2025-08-01     Mumbai           30        85                 12   
1  2025-08-01      Delhi           36        40                 18   
2  2025-08-01  Bengaluru           27        70                 10   
3  2025-08-02     Mumbai           29        88                 14   
4  2025-08-02      Delhi           35        45                 20   
5  2025-08-02  Bengaluru           26        72                  9   
6  2025-08-03     Mumbai           28        90                 11   
7  2025-08-03      Delhi           34        50                 22   
8  2025-08-03  Bengaluru           27        68                  8   

  Weather Condition  
0             Rainy  
1             Sunny  
2            Cloudy  
3             Rainy  
4             Sunny  
5            Cloudy  
6             Rainy  
7     Partly Cloudy  
8            Cloudy  

DataFrame after applymap (numeric values +1):


/tmp/ipython-input-4094409455.py:8: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df[df_numeric.columns]=df_numeric.applymap(lambda x:x + 1)


Date       City  Temperature  Humidity  Wind Speed (km/h)  \
0  2025-08-01     Mumbai           31        86                 13   
1  2025-08-01      Delhi           37        41                 19   
2  2025-08-01  Bengaluru           28        71                 11   
3  2025-08-02     Mumbai           30        89                 15   
4  2025-08-02      Delhi           36        46                 21   
5  2025-08-02  Bengaluru           27        73                 10   
6  2025-08-03     Mumbai           29        91                 12   
7  2025-08-03      Delhi           35        51                 23   
8  2025-08-03  Bengaluru           28        69                  9   

  Weather Condition  
0             Rainy  
1             Sunny  
2            Cloudy  
3             Rainy  
4             Sunny  
5            Cloudy  
6             Rainy  
7     Partly Cloudy  
8            Cloudy

# THANKYOU